### DISCUSSION TITLE
Connector to Other Languages

### DISCUSSION THREAD

**Discussion by stellarpower** [nushell/nushell#10886] [General]
I was wondering - is there any way to get access to data from a nushell session via an API? Would I be able to use the plugin system for this?

I have been looking for a better shell for quite some time - fish has (thus far) unparalleled interactive use - colours, tab completions, etc. - but it's still POSIX-like and processes text, so I don't write anything but trivial sysadmin scripts and wrappers of a few lines. However, the other languages I know well and am comfortable using for more advanced tasks lack the ability to call into system programs well. I don't like Python to start with - but the Xonsh syntax constantly switching between shell mode and Python mode was very ugly. 

Now I am at a bit of a sticking point - because "thinking in nu" is not coming naturally to me and is slowing my work down. But it is much better for working in the terminal than going back to a scripting language and having to constantly work with stdout and stdin when interfacing with the system at large.

I am wondering - what about the ability to patch nu into other languages to do some dogwork for it? If I could send e.g. a table into a different language, where I can perform other data processing - or let's say, use a library, that I can't use from nu - then nu is handling the interactive element, but when someone goes beyond what is easy in nu alone, something else is there able to take the strain - ideally transparently. Nu collects structured data, checks the types, sends this to another language or library, and gets structured data back. A bit like a form of connector or RPC mechanism. I just send a table, get back a list, whatever, same as other languages wrap e.g. C libraries with FFI. 

It'd be fab to be able to open up access to all sorts of other APIs that are only available in a different language and interact with these straight from the shell using nu. googling there seem to be a couple of mentions of something for Python, so I guess this must be possible to an extent already.

Thanks!

**Comment by fdncred** [comment]
The easiest way to exchange nushell's output to another language to consume is with `to json`. Just end any pipeline with `| to json` and nushell should transform the tablular output to normal json for easy consumption.

The opposite is also true. Any external program that can output json can be read with nushell with `| from json` or `| from json -o` if it's one json per line/ndjson/jsonl.

**Reply by stellarpower** [reply]
Agree that that's the easiest way - the problem there is then I'm back to parsing things from text, which partly defeats the point of using nu, as it does this really well. I am thinking more about getting access to the data in-memory.

The example from [this python SDK]( https://github.com/presidento/nushell-python-sdk  )   didn't run for me on my present version, but I think it's closer to what I was thinking.

It'd be neat to be able to write small custom commands in a different language, and just ping and pong data back and forth - or even better if it can be done without making a copy. E.g. if a DataFrame uses Polars - then if another language has polars bindings, I could just reach in and manipulate data that way, and return a new or mutated DataFrame.

**Reply by fdncred** [reply]
The python plugin doesn't work. Someone familiar with Python and nushell will hopefully fix it. Having said that, plugins can be written in any language that supports json or msgpack. Our json used to be json-rpc but it's just json now. There was one in c# but it's using an outdated version of nushell like 0.32.0.

**Reply by stellarpower** [reply]
I see Ruby can consume crates, so if I had the time I'd be intrigued how far that would go. But figured best start with the Python to give it a try. It doesn't look particularly maintained, but at only 6 months old so interesting it's already broken.
